### ISSUE TITLE
Example config for MP Metrics feature

### ISSUE THREAD

**Issue by Charlotte-Holt** [OpenLiberty/docs#1128]
**SME**: Don Bourne

**Resources**:
- https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_mp_metrics_monitor.html 
- https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/rwlp_mp_metrics_rest_api.html

**Comment by Rwalls1** [comment]
#1233

**Comment by Rwalls1** [comment]
#1235

**Comment by Rwalls1** [comment]
Link to current draft: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#mpMetrics-2.2.html

**Comment by Rwalls1** [comment]
@donbourne  I have the following topic ready for review: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#mpMetrics-2.2.html. This draft is intended to add relevant config snippets examples that help to understand the configuration of the feature. Please review to verify whether the included examples are relevant to the feature and suggest any examples that should be added. Also,  If the "Metrics endpoint formats" section is relevant, what example/examples do you think might best illustrate the relevancy of the JSON output format?

**Comment by lauracowen** [comment]
I think this topic needs some work. I don't think we clearly outlined what it should contain but the draft topic currently only partially addresses the requirement for the content in this type of topic.

- Take a look at https://draft-openlibertyio.mybluemix.net/docs/ref/general/#metrics-catalog.html and https://draft-openlibertyio.mybluemix.net/docs/ref/general/#microservice_observability_metrics.html for background about MP metrics and observability of apps on Open Liberty.
- The first (generated) sentence of this topic, while accurate is not that helpful. We can get that fixed in the runtime later but, for now, it would be useful if you could create a `description.adoc` file to include into this topic (in the same way as you have the `examples.adoc` file included at the moment) then in the `description.adoc` file put something like this text (which I've pinched from Charlotte's metrics concept topic):

```
Microprofile Metrics provides a `/metrics` endpoint from which you can access all metrics emitted by the Open Liberty server and deployed applications.
When the application is running, you can view your metrics from any browser by visiting the endpoint. Operations teams can gather the metrics and store them in a database by using tools like Prometheus.
They can then visualize the metrics in dashboards, such as Grafana, to analyze the data.
```
That will then be displayed straight after that first generated sentence to give much more useful intro to the feature. With that, the first sentence isn't so bad (feel free to shout if you think the first sentence is a bad summary of the feature, though - we can definitely get this fixed - I've fixed a couple of others, it just requires a PR to the runtime code repo).
- The information in the "REST endpoints" and "Metrics endpoint formats" sections is not relevant to config examples, which is what should go in this topic. I know this info is taken from one of the two KC links provided but that topic doesn't contain any config examples. The KC topic does potentially contain useful reference information but it is more appropriate elsewhere; depending on the target audience in each case, it should go in one of our other existing Metrics-related topics (https://draft-openlibertyio.mybluemix.net/docs/ref/general/#observability-monitoring.html, https://draft-openlibertyio.mybluemix.net/docs/ref/general/#metrics-catalog.html, https://draft-openlibertyio.mybluemix.net/docs/ref/general/#microservice_observability_metrics.html) or perhaps even elsewhere. I think it's worth talking @Charlotte-Holt  about this because she's worked on similar material already and there might be a reason why this information hasn't already been included anywhere else. If this info is useful and not included elsewhere already, can you and she work with Don to identify which persona/audience it is most useful to and where it should go (if anywhere)? Thanks
- The "Monitoring with MP metrics config" and "MicroProfile metrics API authorization" sections both address the need to secure your metrics endpoint.
  - Start with a sub-heading of something like "Securing the /metrics endpoint". Give a brief explanation about how it's important that not just anyone can access the metrics produced by your app, server, and general environment. Then find out what the _most minimal common_ example is that would be required to get this to work and provide that config example. Check with the SME whether it is necessary to enter anything in the server.xml or whether a basic keystore etc is generated.
- Then have a section called something like "Authorizing access to the metrics API", or maybe split into two similar sections, one in which a user registry is used, and one in which an SSO service is used. Give config examples for each (similar to what's in the KC topic but more minimal - talk to the SME about what is most appropriate). For the SSO version of the config, it might be that the JWT examples shown at the end of the KC topic is useful - ask the SME.
- (check this with the SME that this is okay to say) Add another example with the sub-heading of something like "Disabling metrics security during development". In that section, give the config example of 
```
<mpMetrics authentication="false"/>
```
Before the example, explain that it's possible to disable during development because that makes it easier etc etc but don't do this in production. Also, check with the SME what the default behaviour is and whether this statement is actually required.

If any of that's not clear, ping me next week and we can chat about it.

**Comment by donbourne** [comment]
Hi @Rwalls1 , I see this is in SME Review state -- but not sure if that was intended yet.  From what I can see at https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#mpMetrics-2.2.html , it lacks changes per @lauracowen 's comments

**Comment by Rwalls1** [comment]
@donbourne I wanted to gain your feedback in addition to Laura's comments. Do you have any further suggestions?

**Comment by donbourne** [comment]
@Rwalls1 , I think the main thing is to get this to truly be a topic on the config examples.  What @lauracowen outlined above seems like a good way to approach it and should help show the different ways to configure it under different circumstances, and how to choose between them.

**Comment by Rwalls1** [comment]
@donbourne In relation to Laura's comments, what would be an effective example that illustrates securing the /metrics endpoint?

**Comment by donbourne** [comment]
@Channyboy , could you please provide config example for @Rwalls1 to show how to configure metrics securely (see above for context)

**Comment by Rwalls1** [comment]
@donbourne Please review, I think I've addressed most of Laura's feedback, @channyboy notified me that the metrics feature secures the endpoint by default so I didn't include a config example for that: 
https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#mpMetrics-2.3.html

**Comment by Rwalls1** [comment]
@lauracowen Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#mpMetrics-2.3.html

**Comment by donbourne** [comment]
- "This feature provides support for the Eclipse MicroProfile Metrics 2.3 specification for enterprise Java."
  - true, but I think we should start by explain why someone would want to use the feature

- "The endpoint can be accessed through the metrics API that enables security configuration."
  - the "that enables security configuration" part doesn't make sense.

- somewhere in the intro we need to describe that the MP metrics feature also can provide access to metrics about server components when it is used in combination with the monitor-1.0 feature

- Monitoring with MicroProfile metrics configuration
  - not clear what this section heading is trying to call out 

- "You can use the MicroProfile metrics feature to secure the REST API with authentication"
-> "You can configure the MicroProfile metrics feature to require authentication and authorization to access the /metrics endpoint"

- "The /metrics endpoint reveals all of the service metrics from the location where the application runs. These metrics can identify important data within the application, such as the details of user requests, that could be exposed to unauthorized users if it is not properly secured."
-> "The /metrics endpoint is used to access the metrics stored in the server. This can include both information about server components, such as connection pools, and information that applications have chosen to create metrics for.  In some cases you may want to limit access to this data."


- "Application developers can add their own custom metrics by using the MicroProfile metrics API alongside the metrics provided by Open Liberty. Access to the MicroProfile metrics API can be provided through either a configured user registry or a specified SSO method such as JSON Web Token (JWT) by configuring the administrator role. With Open Liberty, you can map the administrator role to users so that they can access the MicroProfile metrics API. The MicroProfile metrics API is protected by the administrator role so users that are not mapped to the administrator role are not able to access the API."
  - this paragraph is somewhat confusing.  would be better to leave all specifics about each method of securing the endpoint to the section specific to that method.

- "Authorizing access to the metrics API with SSO" 
  - doesn't seem like a complete config example.  eg. missing info about what features need to be enabled

- "Disabling metrics security during development" / "it is possible to disable metrics security during development"
  - there can be cases where you don't care about security in production

**Comment by Rwalls1** [comment]
@donbourne Please review, I think I've addressed most of your feedback, the first line is auto generated so I'll need to work with Laura to update. Also, for the "Authorizing access to the metrics API with SSO" config example, it is actually a complete example that was pulled from https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_mp_metrics_monitor.html. Maybe we can add a line to the parapraph that mentions the enabled feature?

**Comment by donbourne** [comment]
comments that still need to be addressed...

- "This feature provides support for the Eclipse MicroProfile Metrics 2.3 specification for enterprise Java."
  - true, but I think we should start by explain why someone would want to use the feature

  
new comments...

- "The endpoint can be accessed through the metrics API." - this can just be removed (we already say "Microprofile Metrics provides a /metrics endpoint from which you can access..." in the same paragraph)

- "authorization to access the /metrics endpoint" - the "/metrics" part should be in code format

- "Securing the /metrics REST endpoint" - can probably remove that line.

- on the SSO question can you please check with @chunlongliang-ibm  on what we should add there?

**Comment by Rwalls1** [comment]
@donbourne Please review, I will  work with Laura to address the first comment. For the SSO question, I spoke with Bruce Tiffany and he indicated that within the context of this topic, the MP Metrics would be enabled, however, any additional features that had a security constraint defined could also be applicable.

**Comment by donbourne** [comment]
@Rwalls1 , looks good!

**Comment by Rwalls1** [comment]
@lauracowen Please review: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#mpMetrics-2.3.html

**Comment by lauracowen** [comment]
- "The MicroProfile metrics feature" - should be capitalised as "The MicroProfile Metrics feature" in each case when referring to the feature by name (there are a few instances). In the first Examples sub-heading, I think it should either be capitalised in the same way, or remove "MicroProfile" (as the "MicroProfile" prefix implies the name of the feature). I think just capitalising and keeping "MicroProfile" it is fine here.
- Similarly, "MicroProfile Metrics API" - it's the name of the API. Though I think when you don't have "MicroProfile" prefix, lower-case is fine (eg in the heading "Authorizing access to the metrics API with a basic user registry" I think it's fine).
- The short paragraph "The `/metrics` endpoint.....limit access to this data." doesn't really apply to this section (I think it's referring to wanting to limit access to certain roles, which is part of the next example). So I think move that paragraph to insert it after the "Authorizing access to..." heading, so it's a new first paragraph before the "To allow a user or group..." paragraph. I know this isn't necessarily how you'd write a normal topic or article etc but in the case of this topic format, I think we can assume readers will skim down the headings or the code snippets looking for the example most relevant to them. So the info needs to be in the relevant sections, rather than linking one section to the next (as that paragraph seems to currently do).
- SSO section - where do the values for the `user/group-access-id`come from?

Other than that, looks good. Thanks

**Comment by Rwalls1** [comment]
@Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#mpMetrics-2.3.html

In regards to the question about the user/group-access id values, here's a response from Bruce Tiffany:
First, in security there is the concept of “realms”,  which are sort of like domains or namespaces.  Users and groups are scoped to realms.  For example,  admins at google is not the same as admins at paypal.  With single sign on, it’s possible to have authentication provided in multiple realms, such as authenticating with a choice of paypal or google.  In that case, liberty needs to know which realm that user or group belongs to.  This is where access-ids come in. A user or a group in access-id form is usually (realm)/name.  When using Json Web Tokens,  a token attribute called the “issuer”, which is usually a URL, is used as the realm.  So in the above example it looks like the issuer was https://idp.example.com.  There’s a bit of an explanation here https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/cwlp_authorization.html#d337475e272 under “Access IDs and authorization”.
It’s a common developer problem to have applications failing with 403 (unauthorized) and then struggle to get the access-id correct in the xml.
When a “local” registry is used, such as quickStartSecurity or basicRegistry, access id’s aren’t needed, regular old groups and users will suffice.

**Comment by lauracowen** [comment]
- Thanks - re realms, maybe under that example, just add a sentence or two that explains that the user `bob` and the group `ManagingGroup` which are authorised in this example belong to the `https://idp.example.com` realm, which uniquely  (if that makes sense!). In these example config things, we need to give explanations of what the values show in the example where relevant. It doesn't take much I think though - certainly don't get into explaining the general concept here, just make sure it's clear what the value comes from so that the the reader knows to insert the relevant value of their own if they copy the example.

Other than that, I'm good. So let me know when you've got that sorted and I'll sign it off.

- On the grounds that developers have problems with that realm thing, can you add it to the list of troubleshooting tips we're compiling here: https://github.com/OpenLiberty/docs/issues/1686

(I'll also talk to David/Charlotte about the bits in that KC topic that aren't in the OL docs and what, if anything, to do about them. And raise some issues on the backlog if necessary.)

**Comment by Rwalls1** [comment]
@lauracowen Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#mpMetrics-2.3.html

**Comment by lauracowen** [comment]
Thanks. I'll sign this off.

Can you make add a note to the issue I mentioned in my last comment though, please? (just add a comment to the issue and paste in the problem that Bruce described). Thx

**Comment by Rwalls1** [comment]
@lauracowen Thanks, I've added a comment to that issue.

**Comment by Rwalls1** [comment]
@Charlotte-Holt Please peer review: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#mpMetrics-2.3.html

**Comment by Charlotte-Holt** [comment]
@Rwalls1 Looks good! 

### Peer review feedback

- "The metrics can include both information about server components, such as connection pools, and information that applications choose to create metrics for." --> "The metrics can include both information about server components, such as connection pools, and metrics information that is created by applications." **The first way is just reading a little odd to me.** 
- "This mapping authorizes access to the MicroProfile metrics API." --> "This mapping authorizes access to the MicroProfile Metrics API."
- In, "The user `bob` and the `ManagedGroup` group both belong to the `https://idp.example.com` realm that is authorized to access the API.", should `ManagedGroup` be `ManagingGroup` from the example? Not sure. 
- Could "Although the default configuration for the MicroProfile Metrics feature is to secure the REST API, it is possible to disable metrics security." be rewritten as "Although the default configuration for the MicroProfile Metrics feature secures the REST API, it is possible to disable metrics security."? 
- The words that are in back-ticks in "When you enable an SSO method, such as JWT, you can map the administrator role to the user or group access-id. This mapping authorizes access to the MicroProfile metrics API." seem off to me. Specifically  the `user` or `group access-id`. I'm not sure `group access-id` is a thing. Maybe it should be "`user-access-id` or -group-access-id`" or "`user` or `groups` access ID".

**Comment by Rwalls1** [comment]
@Charlotte-Holt Thanks, I've addressed your feedback

**Comment by Rwalls1** [comment]
New link to topic: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/mpMetrics-2.3.html

**Comment by Rwalls1** [comment]
PR: https://github.com/OpenLiberty/docs/pull/1733/files

**Comment by Rwalls1** [comment]
New link to topic: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/reference/feature/mpMetrics-2.3.html

**Comment by Rwalls1** [comment]
Editing reference: https://learning.oreilly.com/library/view/developing-quality-technical/9780133119046/ch06.html
I used this chapter as a guide to ensure that the description of the config examples were clear and avoided ambiguity. The tips related to avoiding wordiness also helped me to review the descriptions to ensure conciseness.

**Comment by chirp1** [comment]
The version of the topic that includes the configuration examples is at https://draft-openlibertyio.mybluemix.net/docs/20.0.0.12/reference/feature/mpMetrics-2.3.html

**Comment by donbourne** [comment]
would like to reference this topic to address a customer question at https://groups.io/g/openliberty/topic/78185136 .  Can this please be published?

**Comment by Rwalls1** [comment]
@donbourne This topic is currently set to be published with the 20.0.0.12 release along with the other Observability topics that are mentioned here: https://github.com/OpenLiberty/docs/issues/3047

**Comment by chirp1** [comment]
@Rwalls1 Hi RIchard, I have a few comments, but the examples look good!
- "Configure MicroProfile Metrics" seems to draw from step 1 of "Set up monitoring with MicroProfile metrics" at https://www.ibm.com/docs/en/was-liberty/base?topic=environment-monitoring-microprofile-metrics. To me, that step is confusing and so I think your "Configure MicroProfile Metrics" could be rewritten. I think that "quickStartSecurity userName="theUser" userPassword="thePassword"" defines the administrator role for the /metrics endpoint if the user wants to use quickStartSecurity. So, maybe that could be in its own section. and then the discussion about the keystore and the default keystore could be in their own section. Other ideas??
- Change "To allow a user or group of users to access the MicroProfile Metrics API,  you can map the administrator role to the user or group from the configured user registry." to something like "To allow users and groups of users access to the MicroProfile Metrics API, you can map the administrator role to users and groups that are in the configured user registry.".
- Change "The following example shows a configuration to enable a user in the configured user registry and AuthorizedGroup to access the API:" to something like "The following example maps the user bob and the group AuthorizedGroup to the administrator role so that they can access the MicroProfile Metrics API.

**Comment by Rwalls1** [comment]
@chirp1 Thanks, I've addressed your feedback:
For the first point related to the  "Configure MicroProfile Metrics" section, after further research, I'm still a bit confused about how to clarify the example so I'm going to reach out to the SME to get a proper explanation before making changes.

**Comment by Rwalls1** [comment]
@chirp1 Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/latest/reference/feature/mpMetrics-2.3.html
 I've discussed the feedback with the SME, they think it would be more useful to include the keyStore configuration and the `quickStartSecurity` config in the same example. I've updated the description to reflect that `quickStartSecurity` is related to administrative role mapping.

**Comment by chirp1** [comment]
@Rwalls1 Hi Richard, the updates look good. Moving to Ready to Publish.

**Comment by Rwalls1** [comment]
Staging: https://staging-openlibertyio.mybluemix.net/docs/21.0.0.8/reference/feature/mpMetrics-3.0.html

**Comment by Rwalls1** [comment]
Updates have been merged to Vnext and will publish with the 21.0.0.8 release.
